### PR TITLE
Add Hello World description to Sphinx docs

### PR DESCRIPTION
**Pull request by Devjiu**
This commit adds Hello world section and describes Triton's programming model and ways to write performant code on this language.

### CODE REVIEW

File: docs/pics/execution_model.png
Comment (etiotto) [inline docs/pics/execution_model.png:1]: We should use Intel terminology rather than NVidia terminology, So:
  - Block -> WorkGroup
  - Warp -> SubGroup

File: docs/getting-started/start.rst [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+.. image :: ../pics/memory-hierarchy.png
+
+*************
+Triton Syntax
+*************
+
+*********
+Semantics
+*********
```
Comment (aregm) [inline docs/getting-started/start.rst:256]: #####################################
"Hello, World!" and the Programming Model
#####################################

********
Glossary
********

* **GPU Kernel** (or **kernel**) is a program or function compiled separately from the main program to be executed on a target GPU.

* **SPMD** stands for Single Program Multiple Data, a paradigm wherein the same kernel (i.e., program) is replicated multiple times on several hardware instances. The same kernel is applied to different subsets of input data generated by dividing by the number of execution threads.

* **Thread**: represents the smallest execution unit on the GPU. Threads within the same block can communicate and synchronize with each other. 

* **Warp** (Nvidia notation) or **SubGroup**(Intel Notation) is a one-dimensional group of threads. The GPU architecture limits the number of threads per subgroup and is typically a multiple of 32.

* **Block** or **Thread Block** or **WorkGroup** - is a two- or three-dimensional group of threads that execute the same code simultaneously on the GPU. Threads within the same block can share data through shared memory and synchronize their execution. 

* **Grid** A grid is a three-dimensional collection of thread blocks. It defines the overall structure of parallel execution on the GPU. Each grid dimension (X, Y, and Z) represents the number of thread blocks along that axis. Threads within different blocks within the same grid can communicate with each other only through global memory.

*********************
Hello World in Triton
*********************

Triton programs are single-threaded, which makes them easier to write, maintain, optimize, and debug. However, this also begs the existence of automatic parallelization mechanisms capable of generating efficient multi-threaded GPU code from high-level specifications of blocked algorithms.

Also, Triton allows the distribution of kernels in 3D logical thread space. Such a distribution is commonly used to map logical threads to data to provide more data locality. In addition, logical threads will be executed with several HW threads. We will start with the simplest case with one logical thread and one kernel and then show how to expand this to achieve better performance and increase HW utilization. 

The simplest case to observe in Triton is a vector sum of 2 vectors. (We will use the default data type as fp32)

Let us start with a dummy implementation. All Triton kernels require ``@triton.jit`` annotation and accept limited argument types. Supported argument types are:

    1. Pointer to data (Object with ``data_ptr()`` and ``dtype()`` - we will use torch.tensor for this purpose)
    2. Python primitive types: int, float, bool, str
    3. Constexprs (``tl.constexpr``)

.. code-block :: python
    import torch
    import triton
    import triton.language as tl
    @triton.jit
    def add_kernel(x_ptr,  # *Pointer* to first input vector.
                   y_ptr,  # *Pointer* to second input vector.
                   output_ptr,  # *Pointer* to output vector.
                   n_elements: tl.constexpr,  # Size of the vector.
                   ):
        
        # There are multiple 'programs' processing different data. We identify which program
        # we are here:
        pid = tl.program_id(axis=0)  # We use a 1D launch grid so axis is 0.
        
        vec_offset = tl.arange(0, n_elements)
        # x_ptr + vec_offset result is a vector of pointers
        x = tl.load(x_ptr + vec_offset)
        y = tl.load(y_ptr + vec_offset)
        
        output = x + y
        
        # Write x + y back to DRAM.
        tl.store(output_ptr + vec_offset, output)
This kernel loads all available data of whole vector with help of ``tl.arange``. Raw call ``tl.load(x_ptr)`` will load only single value. ``tl.arange`` as many other Triton.language operations operates with ``tl.constexpr`` only. 

So what is ``tl.constexpr``? It is a kernel argument specifier that tells the kernel that its value can be calculated during compilation. In simple words, it should be a static constant for the kernel. If its value changes, the kernel requires recompilation to produce a different executable binary. 

To run this kernel, we will need some wrapper code:

.. code-block :: python 
    def add(x: torch.Tensor, y: torch.Tensor):
        # We need to preallocate the output.
        output = torch.empty_like(x)
        assert x.is_xpu and y.is_xpu and output.is_xpu
        n_elements = output.numel()
        
        # The SPMD launch grid denotes the number of kernel instances that run in parallel.
        # It is analogous to CUDA launch grids. 
        # It can be either Tuple[int], or Callable(metaparameters) -> Tuple[int].
        # In this case, we use a 1D grid where the size is the number of blocks:
        grid = (1, ) # same as (1, 1)
        
        # NOTE:
        #  - Each torch.tensor object is implicitly converted into a pointer to its first element.
        #  - `Triton.jit`'-ed functions can be indexed with a launch grid 
        #  - Don't forget to pass meta-parameters as keyword arguments.
        add_kernel[grid](x, y, output, n_elements=n_elements)
        
        return output
    # %%
    # We can now use the above function to compute the element-wise sum of two `torch.tensor` objects 
    # and test its correctness:
    torch.manual_seed(0)
    size = 1024
    x = torch.rand(size, device='xpu')
    y = torch.rand(size, device='xpu')
    output_torch = x + y
    output_triton = add(x, y)
    print("Torch  output: ", output_torch)
    print("Triton output: ", output_triton)
    print(f'The maximum difference between torch and triton is '
        f'{torch.max(torch.abs(output_torch - output_triton))}')
We should pay attention to the grid in the wrapper code, a Tuple of 2 integers. It describes the amount of kernels that will run on GPU. It describes logical thread space for kernels in terms of blocks like the CUDA programming model. A simple visualization of the kernel's grid is:

.. image :: ../pics/execution_model.png
In the current example, we are using a single kernel as we wrote it without any actual usage of pid. Grid can also be callable and return a tuple on the basis of meta parameters. This approach will be covered later.  

This code has several obvious problems:

    1. Works **only with shapes that are** :math:`2^n` as it used in load operation. 
    2. Don't fully utilize the possibility to use more logical threads. 

So, let's rewrite it to support any shape. We will use the additional parameter ``BLOCK_SIZE`` to split (or tile) the custom shape according to the suitable granularity of the target GPU. 

.. code-block :: python
    @triton.jit
    def add_kernel_fine_grained(x_ptr,  # *Pointer* to first input vector.
                                y_ptr,  # *Pointer* to second input vector.
                                output_ptr,  # *Pointer* to output vector.
                                n_elements: tl.constexpr,  # Size of the vector.
                                BLOCK_SIZE: tl.constexpr,  
                                # Number of elements suitable for load/store granularity. 
                            ):
        
        # There are multiple 'programs' processing different data. We identify which program
        # we are here:
        pid = tl.program_id(axis=0)  # We use a 1D launch grid so axis is 0.
        vec_offset = tl.arange(0, BLOCK_SIZE)
        for i in range(0, tl.cdiv(n_elements, BLOCK_SIZE)):
            block_start = i * BLOCK_SIZE
            vec_block_offset = block_start + vec_offset
            # Mask will be vector of booleans
            # Used to avoid loading rubbish into the last tile. 
            mask = vec_block_offset < n_elements
            
            x_tile = tl.load(x_ptr + vec_block_offset, mask=mask)
            y_tile = tl.load(y_ptr + vec_block_offset, mask=mask)
            output_tile = x_tile + y_tile
            
            tl.store(output_ptr + vec_block_offset, output_tile, mask=mask)
Nice! Now, we have resolved the first problem and can work with arbitrary shapes/sizes of the vectors. Let us optimize it to utilize more threads in our grid. Also, we can remove the loop out of the kernel and convert it into the 1D grid.  

.. code-block :: python
    import torch
    import triton
    import triton.language as tl
    @triton.jit
    def add_kernel(x_ptr,  # *Pointer* to first input vector.
                   y_ptr,  # *Pointer* to second input vector.
                   output_ptr,  # *Pointer* to output vector.
                   n_elements,  # Size of the vector.
                   BLOCK_SIZE: tl.constexpr,  # Number of elements each program should process.
                   ):
        
        # There are multiple 'programs' processing different data. We identify which program
        # we are here:
        pid = tl.program_id(axis=0)  # We use a 1D launch grid so axis is 0.
        
        # This program will process inputs that are offset from the initial data.
        # For instance, if you had a vector of length 256 and block_size of 64, the programs
        # would each access the elements [0:64, 64:128, 128:192, 192:256].
        # Note that offsets is a list of pointers:
        block_start = pid * BLOCK_SIZE
        
        offsets = block_start + tl.arange(0, BLOCK_SIZE)
        # Create a mask to guard memory operations against out-of-bounds accesses.
        mask = offsets < n_elements
        
        # Load x and y from DRAM, masking out any extra elements in case the input is not a
        # multiple of the block size.
        x = tl.load(x_ptr + offsets, mask=mask)
        y = tl.load(y_ptr + offsets, mask=mask)
        
        output = x + y
        
        # Write x + y back to DRAM.
        tl.store(output_ptr + offsets, output, mask=mask)
And apply corresponding changes to wrapper:

.. code-block :: python
    def add(x: torch.Tensor, y: torch.Tensor):
        # We need to preallocate the output.
        output = torch.empty_like(x)
        assert x.is_xpu and y.is_xpu and output.is_xpu
        n_elements = output.numel()
        
        # The SPMD launch grid denotes the number of kernel instances that run in parallel.
        # It is analogous to CUDA launch grids. 
        #
        # Grid can be either Tuple[int], or Callable(metaparameters) -> Tuple[int].
        # In this case, we use a 1D grid where the size is the number of blocks.
        grid = lambda meta: (triton.cdiv(n_elements, meta['BLOCK_SIZE']), )
        
        # NOTE:
        #  - Each torch.tensor object is implicitly converted into a pointer to its first element.
        #  - `Triton.jit`'ed functions can be indexed with a launch grid
        #  - Don't forget to pass meta-parameters as keywords arguments.
        add_kernel[grid](x, y, output, n_elements, BLOCK_SIZE=1024)
        return output
    torch.manual_seed(0)
    size = 98432
    x = torch.rand(size, device='xpu')
    y = torch.rand(size, device='xpu')
    output_torch = x + y
    output_triton = add(x, y)
    print(output_torch)
    print(output_triton)
    print(f'The maximum difference between torch and triton is '
        f'{torch.max(torch.abs(output_torch - output_triton))}')
    
Let's sum up and point out what can be done in the kernel body:

    1. Several python primitive ops: 

        * ``+``, ``-``, ``*``, ``/``, ``//``, ``%``, ``&``, ``>>``, ``<<``, ``<``, ``>``, ``|``, ``^``
        * ``range``, ``min``, ``max``, ``print``, ``float``, ``int``, 
        * ``isinstance``, ``getattr``, ``len`` (need some examples for isinstance, getattr, len)
        * ``for``, ``while``, ``pass`` and literals/constants

    2. tl ops - can be found `here <python-api/triton.language.html>`_
    3. arguments of kernel
    4. call other jit-ed functions

.. image :: ../pics/memory-hierarchy.png
*************
Triton Syntax
*************

*********
Semantics
*********

Comment (aregm) [reply docs/getting-started/start.rst:256]: Fixed the English, please update the commit.

**Review by aregm — CHANGES_REQUESTED** [review]

File: docs/pics/execution_model.png
Comment (Devjiu) [reply docs/pics/execution_model.png:1]: triton.Config uses num_warps parameter. Using terms in documentation that are not used in the code is misleading. All examples use Block_Size. I am describing what triton is now, not what it will be in the future.

Comment (etiotto) [reply docs/pics/execution_model.png:1]: I understand, however this repository is for the Intel port of Triton so it would make sense to clarify the terminology applicable to Intel GPUs.  How about we clarify how the NVidia terminology translates to Intel terminology. Similar to what is done for AMD GPUs. Example:

```
* **Warp** (Nvidia notation) or **wavefront** (AMD notation) is a one-dimensional group of threads. This is a smaller unit for extra grain. The number of threads per block is limited by the GPU architecture and is typically a multiple of 32.
* ```

Comment (Devjiu) [reply docs/pics/execution_model.png:1]: Ok, agree, np

File: docs/getting-started/start.rst (hunk 1 above)
Comment (Devjiu) [reply docs/getting-started/start.rst:266]: updated, thanks

**Comment by etiotto** [comment]
To fix pre-commit issues run ` python3 -m pre_commit run --show-diff-on-failure --color=always --all-files --verbose` and check in any files that have been changed by that command.

File: docs/getting-started/start.rst [hunk 2]
```diff
@@ -0,0 +1,613 @@
...
+
+    @triton.jit
+    def matmul_kernel_dummy(
+            # Pointers to matrices
+            a_ptr, b_ptr, c_ptr,
+            # Matrix dimensions
+            M, N, K,
+            # The stride variables represent how much to increase the ptr by when moving by 1
+            # element in a particular dimension. E.g. `stride_am` is how much to increase `a_ptr`
+            # by to get the element one row down (A has M rows).
```
Comment (avladimi) [inline docs/getting-started/start.rst:310]:
```suggestion
            # to access the element one row down (A has M rows).
```

File: docs/getting-started/start.rst [hunk 3]
```diff
@@ -0,0 +1,613 @@
...
+        matmul_kernel_dummy[grid](
+            a, b, c,  #
+            M, N, K,  #
+            a.stride(0), a.stride(1),  #
+            b.stride(0), b.stride(1),  #
+            c.stride(0), c.stride(1),  #
+        )
+        return c
+
+As you see work with matrixes requires manipulation with pointers and assume specific order of data. It can be inconvenient in some scenarios, and force for additional data movement, that can be done during compilation. To solve this problem, use ``block_ptr``. ``block_ptr`` makes it easy to iterate over multi-dimensional blocks in a larger matrix along different dimensions.
```
Comment (avladimi) [inline docs/getting-started/start.rst:377]:
```suggestion
Working with matrices often involves manipulating pointers and assuming a specific data order. This can be inconvenient in certain scenarios and forces additional data movement that could be done during compilation. To solve this problem, use ``block_ptr``. ``block_ptr`` simplifies iterating over multi-dimensional blocks within a larger matrix along various dimensions. 
```

File: docs/getting-started/start.rst [hunk 4]
```diff
@@ -0,0 +1,613 @@
...
+
+        matmul_kernel[grid](
+            a_ptr=a, b_ptr=b, c_ptr=c,  #
+            M=M, N=N, K=K,  #
+            stride_am=a.stride(0), stride_ak=a.stride(1),  #
+            stride_bk=b.stride(0), stride_bn=b.stride(1),  #
+            stride_cm=c.stride(0), stride_cn=c.stride(1))
+        return c
+
+If another kernel is called within the body of this kernel Triton compiler will try to optimize it and will try to remove unnecessary stores to global DRAM memory. In common case function will be inlined. Therefore, in complex use cases where the location of the data is important and can be changed at runtime, it is better to avoid using other utility cores with load/store inside them.
```
Comment (avladimi) [inline docs/getting-started/start.rst:442]:
```suggestion
If another kernel is called within the body of this kernel, the Triton compiler will attempt to optimize it by eliminating unnecessary stores to global DRAM memory. Typically, the function will be inlined. Therefor, in complex use cases where the location of the data is important and can be changed at runtime, avoid using other utility cores with load/store operations inside them. 
```

File: docs/getting-started/start.rst [hunk 5]
```diff
@@ -0,0 +1,613 @@
...
+            stride_cm=c.stride(0), stride_cn=c.stride(1))
+        return c
+
+If another kernel is called within the body of this kernel Triton compiler will try to optimize it and will try to remove unnecessary stores to global DRAM memory. In common case function will be inlined. Therefore, in complex use cases where the location of the data is important and can be changed at runtime, it is better to avoid using other utility cores with load/store inside them. 
+
+********************
+Triton Kernel Tuning
+********************
+
+Triton has builtin possibility to chose between several predefined configurations by running them. It's supported with ``@triton.autotune(configs=[...])`` decorator.
```
Comment (avladimi) [inline docs/getting-started/start.rst:448]:
```suggestion
Triton has a built-in feature that allows you to choose between several predefined configurations by executing them. This functionality is supported by using the ``@triton.autotune(configs=[...])`` decorator. 
```

File: docs/getting-started/start.rst [hunk 6]
```diff
@@ -0,0 +1,613 @@
...
+        triton.Config(meta={'BLOCK_SIZE': 1024}, num_warps=8),
+    ],
+    key=['x_size'] # the two above configs will be evaluated anytime
+                    # the value of x_size changes
+    )
+    @triton.jit
+    def kernel(x_ptr, x_size, **META):
+        BLOCK_SIZE = META['BLOCK_SIZE']
+
+``triton.Config`` describes several parameters for kernel. Most important and commonly used parameter is meta, that is a dictionary of meta-parameters to pass to the kernel as keyword arguments. And all ``tl.constexpr`` arguments of kernel can be passed through meta. Example of usage autotuning with several configs is:
```
Comment (avladimi) [inline docs/getting-started/start.rst:463]:
```suggestion
``triton.Config`` describes several parameters for a kernel. The most important and commonly used parameter is ``meta``, which is a dictionary of meta-parameters passed to the kernel as keyword arguments. Additionally, all ``tl.constexpr`` arguments of the kernel can be passed through ``meta``. Here is an example demonstrating the usage of autotuning with multiple configurations: 
```

File: docs/getting-started/start.rst [hunk 7]
```diff
@@ -0,0 +1,613 @@
...
+        ],
+        key=['M', 'N', 'K'],
+    )
+    @triton.jit
+    def matmul_kernel_dummy(
+            # Pointers to matrices
+            a_ptr, b_ptr, c_ptr,
+            # Matrix dimensions
+            M, N, K,
+            # The stride variables represent how much to increase the ptr by when moving by 1
```
Comment (avladimi) [inline docs/getting-started/start.rst:495]:
```suggestion
            # The stride variables represent how much to increase the pointer by when moving by one
```

**Comment by vlad-penkin** [comment]
@Devjiu what is the current status of this PR? 

Could please create the issue and link it to this PR.

**Comment by Devjiu** [comment]
The failure of integration tests appears to be unrelated to changes in this pr.

**Review by aregm — APPROVED** [review]
